daemon/k8s_watcher.go

```diff
@@ -1550,26 +1623,88 @@ func (d *Daemon) addCiliumNetworkPolicyV2(ciliumV2Store cache.Store, cnp *cilium
 		}
 	}
 
-	var cnpns cilium_v2.CiliumNetworkPolicyNodeStatus
 	if err != nil {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			OK:          false,
-			Error:       fmt.Sprintf("%s", err),
-			LastUpdated: cilium_v2.NewTimestamp(),
-		}
 		scopedLog.WithError(err).Warn("Unable to add CiliumNetworkPolicy")
 	} else {
-		cnpns = cilium_v2.CiliumNetworkPolicyNodeStatus{
-			OK:          true,
-			LastUpdated: cilium_v2.NewTimestamp(),
-		}
 		scopedLog.Info("Imported CiliumNetworkPolicy")
 	}
 
-	go func() {
-		k8s.UpdateCNPStatusV2(ciliumNPClient.CiliumV2(), ciliumV2Store,
-			k8s.BackOffLoopTimeout, node.GetName(), cnp, cnpns)
-	}()
+	ctrlName := cnp.GetControllerName()
+	cnpCM.UpdateController(ctrlName,
+		controller.ControllerParams{
+			DoFunc: func() error {
+				cnpns := cilium_v2.CiliumNetworkPolicyNodeStatus{
+					OK: true,
+				}
+
+				d.policy.Mutex.RLock()
+				rev := d.policy.GetRevision()
+				d.policy.Mutex.RUnlock()
+
+				cnpns.Revision = rev
+
+				ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
+				defer cancel()
+
+				waitForEPsErr := endpointmanager.WaitForEndpointsAtPolicyRev(ctx, rev)
+
+				// If the deadline was reached then not all endpoints are enforcing
+				// the given policy.
+				cnpns.Enforcing = waitForEPsErr == nil
+
+				serverRuleStore, exists, err := ciliumV2Store.Get(cnp)
+				if !exists {
+					// If the CNP doesn't exist we clean up the controller
+					cnpCM.RemoveController(ctrlName)
```

Comment:
@tgraf Is removing a controller inside a function that runs in the controller acceptable?

Reply:
It wouldn't crash, but note that the remove doesn't actually force it to stop. It removes it from the manager's map and closes a stop channel. If the controller is blocked in a DoFunc it won't ever actually return :(

Reply:
`go cnpCM.RemoveController(ctrlName)`?

Reply:
hmm, we should probably check the uuid of the controller we're removing from the map in the removeController calls.

Reply:
Please re-check the code